### PR TITLE
Buy product

### PR DESCRIPTION
**Pull request by kmatlaha**
Buy product

### CODE REVIEW

File: tests/buyProducts_test.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    let totalPrice = await checkoutPage.getTotalPrice();
+    console.log(totalPrice);
+    checkoutPage.clickConfirmOrderButton();
+    checkoutPage.successfulOrderTextCheck();
+    I.assertEqual(price+priceWithColor+flatShippingRatePrice+ecoTaxPrice+vatPrice,totalPrice,'not equal');
+    checkoutPage.clickMyAccountButton();
+    checkoutPage.clickOrderHistoryButton();
+    let lastOrderID = await checkoutPage.getOrderIdText();
+    console.log(lastOrderID);
+    pause();
```
Comment (adudyak) [inline tests/buyProducts_test.js]: remove pause

File: tests/buyProducts_test.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    checkoutPage.verifyCheckoutPageText();
+    checkoutPage.newBillingAddressButtonClick();
+    checkoutPage.fillCheckoutDetails(billingDetails);
+    checkoutPage.clickregionStateButton();
+    checkoutPage.clickregionSelectorChoice();
+    checkoutPage.clickContinueButtonOnCheckoutPage();
+    checkoutPage.clickContinueForDeliveryDetails();
+    checkoutPage.clickContinueDeliveryMethodButton();
+    checkoutPage.clickTermsAndConditionsAgreementButton();
+    checkoutPage.clickContinueForPaymentMethod();
```
Comment (adudyak) [inline tests/buyProducts_test.js]: decrease amount of functions called in scenario from checkoutPage by creating more generic one like `comleteSteps2-5`

File: steps_file.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      this.click(signInButton);
+      this.fillField(emailField, user.email);
+      this.fillField(passwordField, user.password);
+      this.click(submitButton);
+
+
+    }
+    // Define custom steps here, use 'this' to access default methods of I.
+    // It is recommended to place a general 'login' function here.
+
```
Comment (adudyak) [inline steps_file.js]: remove comments and empty new lines

File: pages/product.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  
+  parsedInNum(str){
+    let check = false , temp = "";
+      for(let i=0; i<str.length; i++) {
+        if(check) temp += str[i];
+        if (str[i] == '$') check = true;
+      }   
+      return str = parseFloat(temp);
+  },
```
Comment (adudyak) [inline pages/product.js]: Use .replaceAll() for cleanup.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll
Google for regex to keep only numbers and dot.

File: pages/product.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    return this.parsedInNum(await I.grabTextFrom(this.priceText));
+  },
+
+  async getProductPriceWithChosenColor() {
+    I.click(this.colorPriceButton);
+    I.click(this.colorChoosePress);
+    return this.parsedInNum(await I.grabTextFrom(this.colorChoosePress));
+  },
+  
+  parsedInNum(str){
```
Comment (adudyak) [inline pages/product.js]: function should be a verb, like `parseStringToFloat()`

File: pages/checkout.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+termsAndConditionsAgreementButton: {xpath: '//*[@id="agree1"]'},
+continueForPaymentMethod: {xpath: '//*[@id="button-payment-method"]'},
+flatShippingRatePriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[2]/td[2]'},
+ecoTaxPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[3]/td[2]'},
+vatPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[4]/td[2]'},
+totalPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[5]/td[2]'},
+confirmOrderButton: {xpath: '//*[@id="button-confirm"]'},
+successfulOrderText: 'Your order has been placed!',
+orderIdText: {xpath: '//*[@id="content"]/div[1]/table/tbody/tr[1]/td[1]'},
+myAccountButton: {xpath: '//*[@id="top-links"]/ul/li/span/span'},
```
Comment (adudyak) [inline pages/checkout.js]: use text as starter point, that will be more unique

File: pages/checkout.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+orderHistoryButton: {xpath: '//*[@id="top-links"]/ul/li/ul/li[2]/a'},
```
Comment (adudyak) [inline pages/checkout.js]: use text as starter point, that will be more unique

File: pages/checkout.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+regionSelectorChoice: {xpath: '//a[contains(text(), "Aberdeen")]'},
+continueOnCheckoutPageButton: {xpath: '//*[@id="button-payment-address"]'},
+continueForDeliveryDetails: {xpath: '//*[@id="button-shipping-address"]'},
+continueDeliveryMethodButton: {xpath: '//*[@id="button-shipping-method"]'},
+termsAndConditionsAgreementButton: {xpath: '//*[@id="agree1"]'},
+continueForPaymentMethod: {xpath: '//*[@id="button-payment-method"]'},
+flatShippingRatePriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[2]/td[2]'},
+ecoTaxPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[3]/td[2]'},
+vatPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[4]/td[2]'},
+totalPriceText: {xpath: '//*[@id="collapse-checkout-confirm"]/div/div[1]/table/tfoot/tr[5]/td[2]'},
```
Comment (adudyak) [inline pages/checkout.js]: aren't there a better locators for rows 20-23?

File: pages/checkout.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+async getVatPrice() {
+  return this.parsedInNum(await I.grabTextFrom(this.vatPriceText));
+},
+
+async getTotalPrice() {
+  return this.parsedInNum(await I.grabTextFrom(this.totalPriceText));
+},
+
+parsedInNum(str){
```
Comment (adudyak) [inline pages/checkout.js]: You have this method in product.js.
You can move it to steps_file.js.
Later we will learn Helpers functionality, which stores such functions in generic place.

**Review by adudyak — APPROVED** [review]

File: tests/buyProducts_test.js (hunk 1 above)
Comment (kmatlaha) [reply tests/buyProducts_test.js]: fixed

File: tests/buyProducts_test.js (hunk 2 above)
Comment (kmatlaha) [reply tests/buyProducts_test.js]: fixed

File: steps_file.js (hunk 3 above)
Comment (kmatlaha) [reply steps_file.js]: fixed

File: pages/product.js (hunk 4 above)
Comment (kmatlaha) [reply pages/product.js]: fixed

File: pages/product.js (hunk 5 above)
Comment (kmatlaha) [reply pages/product.js]: fixed

File: pages/checkout.js (hunk 6 above)
Comment (kmatlaha) [reply pages/checkout.js]: fixed

File: pages/checkout.js (hunk 7 above)
Comment (kmatlaha) [reply pages/checkout.js]: fixed

File: pages/checkout.js (hunk 9 above)
Comment (kmatlaha) [reply pages/checkout.js]: fixed

File: tests/buyProducts_test.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    I.login(loginUser);
+    I.see('My Account');
+    I.amOnPage('http://opencart.qatestlab.net/index.php?route=product/product&path=32&product_id=43');
+    let price = await productPage.getProductPrice();
+    console.log(price);
+    let priceWithColor = await productPage.getProductPriceWithChosenColor();
+    console.log(priceWithColor);
+    productPage.addProductToCart();
+    checkoutPage.verifyCheckoutPageText();
+    checkoutPage.clickButton(newBillingAddressButton);
```
Comment (adudyak) [inline tests/buyProducts_test.js]: Tests fails here because input parameter is not defined.

File: tests/buyProducts_test.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    checkoutPage.clickButtonsByStepsOnCheckoutForm();
+    let flatShippingRatePrice = await checkoutPage.getFlatShippingRatePrice();
+    console.log(flatShippingRatePrice);
+    let ecoTaxPrice = await checkoutPage.getEcoTaxPrice();
+    console.log(ecoTaxPrice);
+    let vatPrice = await checkoutPage.getVatPrice();
+    console.log(vatPrice);
+    let totalPrice = await checkoutPage.getTotalPrice();
+    console.log(totalPrice);
+    checkoutPage.clickButton(confirmOrderButtonOnCheckoutPage);
```
Comment (adudyak) [inline tests/buyProducts_test.js]: and most likely will fail here

File: tests/buyProducts_test.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    checkoutPage.successfulOrderTextCheck();
+    I.assertEqual(price + priceWithColor + flatShippingRatePrice + ecoTaxPrice + vatPrice, totalPrice, 'not equal');
+    checkoutPage.idOfLastOrderCheck();
```
Comment (adudyak) [inline tests/buyProducts_test.js]: create separate orderHistory page object. Checkout page does not have list of orders

File: pages/checkout.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+  async getTotalPrice() {
+    return I.parseStringToFloat(await I.grabTextFrom(this.totalPriceText));
+  },
+
+  successfulOrderTextCheck() {
+    I.see(this.successfulOrderTextForSuccess);
+  },
+
+  idOfLastOrderCheck() {
```
Comment (adudyak) [inline pages/checkout.js]: function should start with a verb. `openOrdersHistory()`
`check...()` functions usually return true or false.

**Review by adudyak — CHANGES_REQUESTED** [review]

File: tests/buyProducts_test.js (hunk 10 above)
Comment (kmatlaha) [reply tests/buyProducts_test.js]: fixed

File: tests/buyProducts_test.js (hunk 11 above)
Comment (kmatlaha) [reply tests/buyProducts_test.js]: fixed

File: tests/buyProducts_test.js (hunk 12 above)
Comment (kmatlaha) [reply tests/buyProducts_test.js]: created.

File: pages/checkout.js (hunk 13 above)
Comment (kmatlaha) [reply pages/checkout.js]: fixed.

File: pages/checkout.js [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    I.click(this.regionStateButton);
+    I.click(this.regionSelectorChoice);
+    I.click(this.continueOnCheckoutPageButton);
+    I.click(this.continueForDeliveryDetails);
+    I.click(this.continueDeliveryMethodButton);
+    I.click(this.termsAndConditionsAgreementButton);
+    I.click(this.continueForPaymentMethod);
+  },
+
+  clickButton(nameOfButton) {
```
Comment (adudyak) [inline pages/checkout.js]: remove this function

File: pages/orderHistory.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+const { I } = inject();
+
+module.exports = {
+
+  orderIdText: { xpath: '//*[@id="content"]/div[1]/table/tbody/tr[1]/td[1]' },
+  myAccountButtonOnTopPanel: { xpath: '//*[@id="top-links"]/ul/li/span/span' },
+  orderHistoryButtonInMyAccountDropdown: { xpath: '//*[@id="top-links"]/ul/li/ul/li[2]/a' },
+
+  clickButton(nameOfButton) {
```
Comment (adudyak) [inline pages/orderHistory.js]: remove this function

File: tests/buyProducts_test.js (hunk 10 above)
Comment (adudyak) [reply tests/buyProducts_test.js]: It is very wierd to import button locator from page object in rows 1-2 and then pass it back to page object.
Rename this method to `clickNewAddress()` and don't pass any parameters.

File: tests/buyProducts_test.js (hunk 11 above)
Comment (adudyak) [reply tests/buyProducts_test.js]: apply same actions, as for `checkoutPage.clickButton(newBillingAddressButton);`

**Review by adudyak — CHANGES_REQUESTED** [review]

File: pages/checkout.js (hunk 14 above)
Comment (kmatlaha) [reply pages/checkout.js]: removed

File: pages/orderHistory.js (hunk 15 above)
Comment (kmatlaha) [reply pages/orderHistory.js]: removed

**Review by adudyak — APPROVED** [review]

**Pull request merged**